src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -589,25 +664,30 @@ - (void)startReadWithHandler:(void (^)(grpc_byte_buffer *))handler {
 // If the call is currently paused, this is a noop. Restarting the call will invoke this
 // method.
 // TODO(jcanizales): Rename to readResponseIfNotPaused.
-- (void)startNextRead {
+- (void)maybeStartNextRead {
   @synchronized(self) {
     if (_state != GRXWriterStateStarted) {
       return;
     }
+    if (_callOptions.enableFlowControl && (_pendingCoreRead || _pendingReceiveNextMessages == 0)) {
+      return;
+    }
+    _pendingCoreRead = YES;
+    _pendingReceiveNextMessages--;
   }
 
   dispatch_async(_callQueue, ^{
     __weak GRPCCall *weakSelf = self;
     [self startReadWithHandler:^(grpc_byte_buffer *message) {
-      if (message == NULL) {
-        // No more messages from the server
-        return;
-      }
       __strong GRPCCall *strongSelf = weakSelf;
       if (strongSelf == nil) {
         grpc_byte_buffer_destroy(message);
         return;
       }
+      if (message == NULL) {
+        // No more messages from the server
+        return;
+      }
       NSData *data = [NSData grpc_dataWithByteBuffer:message];
       grpc_byte_buffer_destroy(message);
       if (!data) {
```

Comment:
Agreed. However I was thinking this problem probably falls into the area of flow control and hoping it could be removed. If the server sent too much memory that the client runs out of memory, it's almost sure that they need flow control procedure to handle the situation better.

Reply:
Were you planning to remove the `if (!data)` block? If yes, what were you planning to replace it with? Error log msg?

Reply:
NSAssert(data != nil);
if (data == nil) {
  // the same as current block
}